src/core/config/config_module.rs

```diff
@@ -15,13 +15,45 @@ use crate::core::rest::{EndpointSet, Unchecked};
 
 /// A wrapper on top of Config that contains all the resolved extensions and
 /// computed values.
-#[derive(Clone, Debug, Default, Setters, MergeRight)]
+#[derive(Clone, Debug, Default, Getters, MergeRight)]
 pub struct ConfigModule {
-    pub config: Config,
-    pub extensions: Extensions,
-    pub input_types: HashSet<String>,
-    pub output_types: HashSet<String>,
-    pub interface_types: HashSet<String>,
+    config: Config,
+    extensions: Extensions,
+    input_types: HashSet<String>,
+    output_types: HashSet<String>,
+    interface_types: HashSet<String>,
+}
+
+impl ConfigModule {
+    // recompute the values of `input_types`, `output_types` and `interface_types`
+    pub fn recompute_types(self) -> Self {
+        let input_types = self.config().input_types();
+        let output_types = self.config().output_types();
+        let interface_types = self.config().interface_types();
+        Self {
+            config: self.config,
+            extensions: self.extensions,
+            input_types,
+            output_types,
+            interface_types,
+        }
+    }
+
+    pub fn get_config_mut(&mut self) -> &mut Config {
+        &mut self.config
+    }
+
+    pub fn get_extensions_mut(&mut self) -> &mut Extensions {
+        &mut self.extensions
+    }
+
+    pub fn into_extensions(self) -> Extensions {
```

Comment:
Can add a `add_extension` that inserts an extension.

Reply:
_into_extensions_ returns ownership of extensions to places that needs ownership of extensions which helps to avoid doing cloning on fields that can avoid cloning, _set_extensions_ can probably be used for places that needs mutable reference. I agree that the name is semantically vague, if you're ok with cloning we can just get rid of it and instead use owned values.

Reply:
Yeah we can use clone. Understand that LLVM will remove unnecessary clones wherever possible https://nnethercote.github.io/perf-book/heap-allocations.html#clone